src/autopas/containers/NeighborListsBuffer.h

```diff
@@ -0,0 +1,148 @@
+/**
+ * @file NeighborListsBuffer.h
+ * @author F. Gratl
+ * @date 24.08.23
+ */
+
+#pragma once
+
+#include <cstddef>
+#include <unordered_map>
+#include <vector>
+
+#include "autopas/utils/ExceptionHandler.h"
+
+namespace autopas {
+/**
+ * Class for manual memory management of neighbor lists.
+ * A internal buffer of lists is kept in memory, never deleted, and only its content cleared and reassigned.
+ *
+ * @tparam Key Preferably something that is cheap to destroy.
+ * @tparam Value Has to be default constructable.
+ * @tparam Hash
+ * @tparam KeyEqual
+ */
+template <class Key, class Value, class Hash = std::hash<Key>, class KeyEqual = std::equal_to<Key>>
+class NeighborListsBuffer {
+ public:
+  /**
+   * Getter for a reference to a neighbor list by index.
+   * @tparam throwIfIndexOutOfBounds
+   * @param index
+   * @return
+   */
+  template <bool throwIfIndexOutOfBounds = false>
+  std::vector<Value> &getNeighborListRef(size_t index) {
+    if constexpr (throwIfIndexOutOfBounds) {
+      if (index > _lastValidListIndex) {
+        autopas::utils::ExceptionHandler::exception(
+            "NeighborListsBuffer::getNeighborListRef() index out of bounds: {} > {}", index, _lastValidListIndex);
+      }
+    }
+    return _neighborLists[index];
+  }
+
+  /**
+   * Getter for a reference to a neighbor list by key.
+   * @tparam throwIfKeyIsUnknown
+   * @param key
+   * @return
+   */
+  template <bool throwIfKeyIsUnknown = false>
+  std::vector<Value> &getNeighborListRef(const Key &key) {
+    if constexpr (throwIfKeyIsUnknown) {
+      if (_keyMap.find(key) == _keyMap.end()) {
+        autopas::utils::ExceptionHandler::exception("NeighborListsBuffer::getNeighborListRef() unknown key: {}", key);
+      }
+    }
+    return _neighborLists[_keyMap[key]];
+  }
+
+  /**
+   * Reserves a neighbor list for use.
+   * Grows the internal buffer if there are no spare lists left.
+   *
+   * @note This function is NOT thread safe.
+   *
+   * @return
+   */
+  size_t addNeighborList() {
```

Comment:
`reserveNeighborList()` would be confusing because it is too close to `reserveNeighborLists()`, which does something fundamentally different.

How about `getNewNeighborList()`? -> e80896580